content/drafts/using-sti-to-evolve-entities.md

```diff
@@ -0,0 +1,244 @@
+
+### ...this seems coupled?
+
+There is a tangent/future post about "how/why is `Task` co-mingled with the other schedule-adjacent features?", but my tldr is that cross-feature cohesion is essentially the raison d'etre (reason for being) for custom software systems, because it lets users, and data, stay highly-contextual & interconnected.
+
+Granted, there's also the tangent of whether this coupling should happen via FKs in a monolith, or APIs in microservices, both of which will face the same v1/v2 migration challenge but with different pros & cons, but, again, that's a separate post. :-)
+
+## Goals
+
+So our goals were to:
+
+1. Leave existing data/projects using Task v1 largely in-place,
+
+   Thankfully both our product team & field ops were onboard with leaving all existing projects on the tasks/schedule v1 feature, as the v1 -> v2 migration for existing data would be painful/extremely expensive/and likely create nonsense/unuseful data anyway. Instead, we'll just let all existing projects run-out on v1.
+
+2. Setup new data/projects using Task v2 (when released),
+3. Have the "Task v1" and "Task v2" features know as little about the other as possible, and
+4. Have rest of the system's `Task`-adjacent features be largely agnostic about the v1/v2 dichotomy
+
+And, of course, minimize both our short-term and long-term pain.
+
+## Auditing Existing Usage
+
+Our first step was to understand how `Task` was used in the system, to understand the impact of our proposed changes.
+
+Given our system is a monolith, this was actually pretty easy to do: all incoming and outgoing connections (i.e. foreign keys from `tasks` to other tables, and foreign keys from other tables to `tasks`) to `Task` are listed in Joist's `TaskOpts` type, which a small sample looks like:
+
+```ts
+export interface TaskOpts {
+  // All primitive columns & enums...
+  durationInDays: number;
+  startDate: Date;
+  isStartNode?: boolean | null;
+  baselineStartDate?: Date | null;
+  internalNote?: string | null;
+  
+  // Also any "outgoing" FKs to other parts of the system
+  internalUser?: InternalUser | InternalUserId | null;
+  tradePartner?: TradePartner | TradePartnerId | null;
+  invoice?: Invoice | null;
+  
+  // Also well as any "incoming" FKs that depend on task
+  billLineItems?: BillLineItem[];
+  homeownerContractDraws?: HomeownerContractDraw[];
+  requiredTaskDocuments?: RequiredTaskDocument[];
+  documents?: Document[];
+  jobLogNotes?: JobLogNote[];
+}
+```
+
+Joist also produces a "reactivity report" that further shows how `Task` is connected to the rest of the system, beyond the immediate foreign keys, via Joist's reactive rules & reactive fields features:
```

Comment:
Question: is this something joist provides out of the box? Seems like we have our own script to read the config and print this `reactions` file?

Reply:
You caught me hand-waving on this :-) but yes, the reactions is still an internal-only script atm.